sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+              }
+            });
+      } else {
+        LOG.info("Not splitting source {} because source is already split.", this);
+        return ImmutableList.of((BoundedSource<KV<K, V>>) this);
+      }
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions po) throws Exception {
+      if (inputSplit == null) {
+        if (inputSplits == null) {
+          computeSplits();
+        }
+        return boundedSourceEstimatedSize;
+      }
+      return inputSplit.getSplit().getLength();
+    }
+
+
+    /**
+     * This is helper function to compute splits. This method will also calculate size of the data
+     * being read. Note : This method is called exactly once, the splits are retrieved and cached
+     * for further use by splitIntoBundles() and getEstimatesSizeBytes().
+     * @throws InterruptedException
+     */
+    @VisibleForTesting
+    void computeSplits() throws IOException, InterruptedException{
+      inputFormatObj = createInputFormat();
+      List<InputSplit> splits =
+          inputFormatObj.getSplits(Job.getInstance(conf.getHadoopConfiguration()));
+      if (splits == null) {
+        throw new IOException(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_GETSPLITS_ERROR_MSG);
+      }
+      if (splits.isEmpty()) {
+        throw new IOException(HadoopInputFormatIOContants.COMPUTESPLITS_EMPTY_SPLITS_ERROR_MSG);
+      }
+      boundedSourceEstimatedSize = 0;
+      inputSplits = new ArrayList<SerializableSplit>();
+      for (InputSplit inputSplit : splits) {
+        if (inputSplit == null) {
+          throw new IOException(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_SPLIT_ERROR_MSG);
+        }
+        boundedSourceEstimatedSize = boundedSourceEstimatedSize + inputSplit.getLength();
+        inputSplits.add(new SerializableSplit(inputSplit));
+      }
+      validateKeyValueClasses();
+    }
+
+    /**
+     * Returns instance of InputFormat set in the configuration.
+     */
+    private InputFormat<?, ?> createInputFormat() throws IOException, InterruptedException {
+      InputFormat<?, ?> inputFormatObj;
+      try {
+        inputFormatObj = (InputFormat<?, ?>) conf.getHadoopConfiguration()
+            .getClass(HadoopInputFormatIOContants.INPUTFORMAT_CLASSNAME, null).newInstance();
+        /**
+         * Sets configuration if provided InputFormat implements Configurable.
+         */
+        if (Configurable.class.isAssignableFrom(inputFormatObj.getClass())) {
+          ((Configurable) inputFormatObj).setConf(conf.getHadoopConfiguration());
+        }
+      } catch (InstantiationException e) {
+        throw new IOException("Unable to create InputFormat : ", e);
+      } catch (IllegalAccessException e) {
+        throw new IOException("Unable to create InputFormat : ", e);
+      }
+      return inputFormatObj;
+    }
+
+    /**
+     * Throws exception is InputFormat key/value class set by the user is different than InputFormat's actual key/value class.
```

Comment:
Ok, Adding the explanation "* Throws exception if you set different InputFormat key or value class than InputFormat's
     * actual key or value class. If you set incorrect key or value class then, it may result in an
     * error like "unexpected extra bytes after decoding" while the decoding process happens. Hence
     * this validation is required."